File: src/test/java/com/tartner/dancehours/domain/danceuser/DanceUserAggregateTest.java
package com.tartner.dancehours.domain.danceuser;

import com.tartner.ExpectException;
import com.tartner.dancehours.DanceHoursId;
import com.tartner.dancehours.domain.danceuser.external.CreateDanceUserCommand;
import com.tartner.dancehours.domain.danceuser.external.DanceUserAggregateQueryModel;
import com.tartner.dancehours.domain.danceuser.external.DanceUserCreatedEvent;
import com.tartner.dancehours.domain.danceuser.external.DanceUserEmailAlreadyExistsException;
import com.tartner.dancehours.domain.danceuser.external.DanceUserIdAlreadyExistsException;
import com.tartner.domain.password.PasswordSetEvent;
import com.tartner.domain.password.TestPasswordHolder;
import com.tartner.utilities.KFixtures;
import org.axonframework.domain.DomainEventStream;
import org.axonframework.test.FixtureConfiguration;
import org.jetbrains.annotations.NotNull;
import org.junit.Before;
import org.junit.Test;

import java.util.HashSet;
import java.util.UUID;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.equalTo;
import static org.mockito.Mockito.*;

//********************************************************************************************
//********************************************************************************************
//********************************************************************************************
// WARNING: Do not convert to kotlin yet, the newGivenWhenThenFixture doesn't work correctly
// or I'm doing something wrong
//********************************************************************************************
//********************************************************************************************
//********************************************************************************************

public class DanceUserAggregateTest {

    private static String UUIDString = "e78320a0-48b9-490a-8b1b-f5bcdcf34995";
    public static final UUID CreateUserId = UUID.fromString(UUIDString);
    public static final String CreateEmail = "bdavisx@yahoo.com";
    public static final String CreateFullName = "Bill Davis";
    public static final String CreatePassword = "notReal";

    private FixtureConfiguration fixture;
    private CreateDanceUserCommand createCommand;
    private DanceUserCreatedEvent createdEvent;
    private DanceUserAggregateQueryModel queryModelMock;
    private PasswordSetEvent passwordSetEvent;

    @Before
    public void setUp() throws Exception {
        //fixture = Fixtures.newGivenWhenThenFixture( DanceUserAggregate.class );
        fixture = KFixtures.INSTANCE.newGivenWhenThenFixture(DanceUserAggregate.class);
        createCommand = createValidCreateCommand();
        createdEvent = createCreatedEventForValidCommand(createCommand);
        queryModelMock = mock(DanceUserAggregateQueryModel.class);
        passwordSetEvent = createPasswordSetEvent();
    }

    @Test
    public void danceUserCreated() throws Exception {
        /* Note: We're going to test the command handler and the aggregate
            together because they act as a unit. The command handler should
            only be the container managed side of the unit, it shouldn't
            be doing anything else. */

        DanceUserAggregate user = new DanceUserAggregate();
        user.create(createCommand, queryModelMock, passwordSetEvent);

        verify(queryModelMock).emailAlreadyExists(CreateEmail);
        verify(queryModelMock).userIdAlreadyExists(buildCreateUserId());

        final DomainEventStream events = user.getUncommittedEvents();
        assertThat(createdEvent, equalTo(events.next().getPayload()));
        assertThat(passwordSetEvent, equalTo(events.next().getPayload()));
    }

    @NotNull
    private DanceHoursId buildCreateUserId() {return DanceHoursId.Default.create(CreateUserId);}

    @Test()
    public void danceUserCreatedDuplicateUserId() throws Exception {
        CreateDanceUserCommand command = createValidCreateCommand();

        DanceUserAggregateQueryModel queryModelMock = mock(DanceUserAggregateQueryModel.class);
        when(queryModelMock.userIdAlreadyExists(buildCreateUserId())).thenReturn(true);

        DanceUserAggregate user = new DanceUserAggregate();
        ExpectException.expectException(DanceUserIdAlreadyExistsException.class, () -> {
            user.create(createCommand, queryModelMock, passwordSetEvent);
        });
    }

    @Test
    public void danceUserCreatedDuplicateEmail() throws Exception {
        CreateDanceUserCommand command = createValidCreateCommand();

        DanceUserAggregateQueryModel queryModelMock =
            mock(DanceUserAggregateQueryModel.class);
        when(queryModelMock.emailAlreadyExists(CreateEmail))
            .thenReturn(true);

        DanceUserAggregate user = new DanceUserAggregate();

        ExpectException.expectException(DanceUserEmailAlreadyExistsException.class,
            () -> user.create(createCommand, queryModelMock, passwordSetEvent));
    }

    private CreateDanceUserCommand createValidCreateCommand() {
        CreateDanceUserCommand command =
            new CreateDanceUserCommand(buildCreateUserId(), CreateEmail,
                CreateFullName, CreatePassword, new HashSet<>());
        return command;
    }

    private DanceUserCreatedEvent createCreatedEventForValidCommand(
        final CreateDanceUserCommand command) {
        DanceUserCreatedEvent event = new DanceUserCreatedEvent(buildCreateUserId(),
            command.getEmail(), command.getFullName(), command.getRoles());
        return event;
    }

    private PasswordSetEvent createPasswordSetEvent() {
        final TestPasswordHolder passwordHolder =
            TestPasswordHolder.Companion.CreateDefaultTest();
        return new PasswordSetEvent(createCommand.getUserId(),
            passwordHolder.getPasswordHash(), passwordHolder.getSalt());
    }

}

File: src/test/java/com/tartner/dancehours/domain/danceuser/DanceUserAggregateTest.kt
package com.tartner.dancehours.domain.danceuser

import com.tartner.dancehours.DanceHoursId
import com.tartner.dancehours.domain.danceuser.external.*
import com.tartner.domain.password.PasswordSetEvent
import com.tartner.domain.password.TestPasswordHolder
import com.tartner.expectException
import com.tartner.utilities.KFixtures
import org.axonframework.test.FixtureConfiguration
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.junit.Before
import org.junit.Test
import org.mockito.Mockito.*
import java.util.*

class DanceUserAggregateTest {

    companion object {
        val UUIDString = "e78320a0-48b9-490a-8b1b-f5bcdcf34995"

        val CreateUserId = UUID.fromString(UUIDString)
        val CreateEmail = "bdavisx@yahoo.com"
        val CreateFullName = "Bill Davis"
        val CreatePassword = "notReal"
    }

    private var fixture : FixtureConfiguration<*>? = KFixtures.newGivenWhenThenFixture(DanceUserAggregate::class.java)
    private var createCommand : CreateDanceUserCommand = createValidCreateCommand()
    private var createdEvent : DanceUserCreatedEvent = createCreatedEventForValidCommand(createCommand)
    private var queryModelMock : DanceUserAggregateQueryModel = mock(DanceUserAggregateQueryModel::class.java)
    private var passwordSetEvent : PasswordSetEvent = createPasswordSetEvent()

    @Before
    @Throws(Exception::class)
    fun setUp() {
    }

    @Test
    @Throws(Exception::class)
    fun danceUserCreated() {
        /* Note: We're going to test the command handler and the aggregate
            together because they act as a unit. The command handler should
            only be the container managed side of the unit, it shouldn't
            be doing anything else. */

        val user = DanceUserAggregate()
        user.create(createCommand, queryModelMock, passwordSetEvent)

        verify<DanceUserAggregateQueryModel>(queryModelMock).emailAlreadyExists(CreateEmail)
        verify<DanceUserAggregateQueryModel>(queryModelMock).userIdAlreadyExists(
            buildCreateUserId())

        val events = user.uncommittedEvents
        assertThat<DanceUserCreatedEvent>(createdEvent, equalTo(events.next().payload))
        assertThat<PasswordSetEvent>(passwordSetEvent, equalTo(events.next().payload))
    }

    private fun buildCreateUserId() : DanceHoursId {
        return DanceHoursId.create(CreateUserId)
    }

    @Test
    @Throws(Exception::class)
    fun danceUserCreatedDuplicateUserId() {
        val command = createValidCreateCommand()

        val queryModelMock = mock(DanceUserAggregateQueryModel::class.java)
        `when`(queryModelMock.userIdAlreadyExists(buildCreateUserId())).thenReturn(true)

        val user = DanceUserAggregate()
        expectException(DanceUserIdAlreadyExistsException::class) {
            user.create(createCommand!!, queryModelMock, passwordSetEvent!!)
        }
    }

    @Test
    @Throws(Exception::class)
    fun danceUserCreatedDuplicateEmail() {
        val command = createValidCreateCommand()

        val queryModelMock = mock(DanceUserAggregateQueryModel::class.java)
        `when`(queryModelMock.emailAlreadyExists(CreateEmail))
            .thenReturn(true)

        val user = DanceUserAggregate()

        expectException(DanceUserEmailAlreadyExistsException::class) {
            user.create(createCommand!!, queryModelMock, passwordSetEvent!!)
        }
    }

    private fun createValidCreateCommand() : CreateDanceUserCommand {
        val command = CreateDanceUserCommand(buildCreateUserId(), CreateEmail,
            CreateFullName, CreatePassword, HashSet<DanceUserRole>())
        return command
    }

    private fun createCreatedEventForValidCommand(
        command : CreateDanceUserCommand) : DanceUserCreatedEvent {
        val event = DanceUserCreatedEvent(buildCreateUserId(),
            command.email, command.fullName, command.roles)
        return event
    }

    private fun createPasswordSetEvent() : PasswordSetEvent {
        val passwordHolder = TestPasswordHolder.CreateDefaultTest()
        return PasswordSetEvent(createCommand!!.userId,
            passwordHolder.passwordHash, passwordHolder.salt)
    }

}
